File: experiments/ts/template/game-of-life.ts
export enum Cell {
    Dead,
    Alive,
}

export type Matrix = Cell[][];

export class GameOfLife {
    #matrix: Matrix;

    constructor(matrix: Matrix) {
        this.#matrix = matrix;
    }

    tick(): void {
        if (this.#matrix.length === 0) {
            return;
        }
        const rows: number = this.#matrix.length;
        const cols: number = this.#matrix[0].length;

        const newMatrix: Matrix = JSON.parse(JSON.stringify(this.#matrix));

        for (let row: number = 0; row < rows; row++) {
            for (let col: number = 0; col < cols; col++) {
                const liveNeighbors: number = this.#countLiveNeighbors(row, col, rows, cols);

                let cell: Cell = this.#matrix[row][col];
                if (cell === Cell.Alive) {
                    if (liveNeighbors < 2 || liveNeighbors > 3) {
                        cell = Cell.Dead;
                    }
                } else {
                    if (liveNeighbors === 3) {
                        cell = Cell.Alive;
                    }
                }

                newMatrix[row][col] = cell;
            }
        }

        this.#matrix = newMatrix;
    }

    #countLiveNeighbors(row: number, col: number, rows: number, cols: number): number {
        let count: number = 0;
        for (let newRow: number = row - 1; newRow <= row + 1; newRow++) {
            for (let newCol: number = col - 1; newCol <= col + 1; newCol++) {
                if (newRow === row && newCol === col) {
                    continue;
                }
                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols) {
                    if (this.#matrix[newRow][newCol] === Cell.Alive) {
                        count++;
                    }
                }
            }
        }
        return count;
    }

    state(): Matrix {
        return this.#matrix;
    }
}

File: experiments/game-of-life/python/attempts/opus46high/1/game_of_life.py
def tick(matrix):
    if not matrix:
        return []

    rows = len(matrix)
    cols = len(matrix[0])

    def count_live_neighbors(r, c):
        count = 0
        for dr, dc in [
            (-1, -1), (-1, 0), (-1, 1),
            (0, -1),           (0, 1),
            (1, -1),  (1, 0),  (1, 1),
        ]:
            nr, nc = r + dr, c + dc
            if 0 <= nr < rows and 0 <= nc < cols:
                if matrix[nr][nc] == 1 or matrix[nr][nc] == "Z":
                    count += 1
        return count

    new_matrix = []
    for r in range(rows):
        new_row = []
        for c in range(cols):
            current = matrix[r][c]
            if current == "Z":
                new_row.append("Z")
            else:
                neighbors = count_live_neighbors(r, c)
                state = 0
                if current == 1 and (neighbors == 2 or neighbors == 3):
                    state = 1
                elif current == 0 and neighbors == 3:
                    state = 1
                new_row.append(state)
        new_matrix.append(new_row)

    return new_matrix